packages/cli/src/ui/contexts/KeypressContext.tsx

```diff
@@ -19,427 +14,733 @@ import {
-  if (m) {
-    const sym = m[1];
-    const name = LEGACY_FUNC_TO_NAME[sym]!;
-    return {
-      key: {
-        name,
-        ctrl: false,
-        meta: false,
-        shift: false,
-        paste: false,
-        sequence: buffer.slice(0, m[0].length),
-        kittyProtocol: false,
-      },
-      length: m[0].length,
-    };
-  }
-
-  return null;
+  return (key: Key | null) => {
+    clearTimeout(timeoutId);
+    bufferer.next(key);
+  };
 }
 
 /**
- * Returns the first index before which we are certain there is no paste marker.
+ * Turns raw data strings into keypress events sent to the provided handler.
+ * Buffers escape sequences until a full sequence is received or
+ * until a timeout occurs.
  */
-function earliestPossiblePasteMarker(data: string): number {
-  // Check data for full start-paste or end-paste markers.
-  const startIndex = data.indexOf(PASTE_MODE_START);
-  const endIndex = data.indexOf(PASTE_MODE_END);
-  if (startIndex !== -1 && endIndex !== -1) {
-    return Math.min(startIndex, endIndex);
-  } else if (startIndex !== -1) {
-    return startIndex;
-  } else if (endIndex !== -1) {
-    return endIndex;
-  }
-
-  // data contains no full start-paste or end-paste.
-  // Check if data ends with a prefix of start-paste or end-paste.
-  const codeLength = PASTE_MODE_START.length;
-  for (let i = Math.min(data.length, codeLength - 1); i > 0; i--) {
-    const candidate = data.slice(data.length - i);
-    if (
-      PASTE_MODE_START.indexOf(candidate) === 0 ||
-      PASTE_MODE_END.indexOf(candidate) === 0
-    ) {
-      return data.length - i;
+function createDataListener(keypressHandler: KeypressHandler) {
+  const parser = emitKeys((key) => keypressHandler(key));
+  parser.next(); // prime the generator so it starts listening.
+
+  let timeoutId: NodeJS.Timeout;
+  return (data: string) => {
+    clearTimeout(timeoutId);
+    for (const char of data) {
+      parser.next(char);
     }
-  }
-  return data.length;
+    timeoutId = setTimeout(() => parser.next(''), ESC_TIMEOUT);
+  };
 }
 
 /**
- * A generator that takes in data chunks and spits out paste-start and
- * paste-end keypresses. All non-paste marker data is passed to passthrough.
+ * Translates raw keypress characters into key events.
+ * Buffers escape sequences until a full sequence is received or
+ * until an empty string is sent to indicate a timeout.
  */
-function* pasteMarkerParser(
-  passthrough: PassThrough,
-  keypressHandler: (_: unknown, key: Key) => void,
+function* emitKeys(
+  keypressHandler: KeypressHandler,
 ): Generator<void, void, string> {
   while (true) {
-    let data = yield;
-    if (data.length === 0) {
-      continue; // we timed out
+    let ch = yield;
+    let sequence = ch;
+    let escaped = false;
+
+    let name = undefined;
+    let ctrl = false;
+    let meta = false;
+    let shift = false;
+    let code = undefined;
+
+    if (ch === ESC) {
+      escaped = true;
+      sequence += ch = yield;
```

Comment:
init: can you avoid the compound  statement and add the 
"ch = yield" to its own line so it is clearer what is happening. for a second I was worried about the second ch == ESC line.